### PR TITLE
prevent possible memory leak in Peer

### LINKED ISSUE #1719: memory leak if remote peer doesn't reply to a ping
When we are sending pings to a Peer that never sends `Pong` back for any reason, the `org.bitcoinj.core.Peer#pendingPings` is getting filled with `PendingPing` instances that are never cleared. (Normally they are cleared only on Pong receival)

Can we throw old `PendingPings` away (add timeout)? Disconnect from such Peer (how)? 

Similar problem seems to be with `org.bitcoinj.core.Peer#pendingTxDownloads` that gets filled if some transactions never arrive(?).

### PR DESCRIPTION
**Pull request by pvyhnal-generalbytes**
- when a peer does not respond to Ping,
- when a peer does not send announced transactions 

fixes #1719 

Please review if this way to disconnect is OK.

I was not sure about the number of pending objects before disconnect. Using too low number may lead to reconnecting too often. I saw a peer with around 15000 pending pings when running for around a day but was not able to reproduce it recently. Another peer had around 2000 pending transactions when running over one night.

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/core/Peer.java [hunk 1]
```diff
@@ -1213,6 +1213,11 @@ protected void processInv(InventoryMessage inv) {
             } else {
                 log.debug("{}: getdata on tx {}", getAddress(), item.hash);
                 getdata.addTransaction(item.hash, vPeerVersionMessage.isWitnessSupported());
+                if (pendingTxDownloads.size() > 100) {
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Peer.java]: I would put these constants into a constant field.

File: core/src/main/java/org/bitcoinj/core/Peer.java [hunk 2]
```diff
@@ -1552,6 +1557,10 @@ private void addPingTimeData(long sample) {
         final VersionMessage ver = vPeerVersionMessage;
         if (!ver.isPingPongSupported())
             throw new ProtocolException("Peer version is too low for measurable pings: " + ver);
+        if (pendingPings.size() > 50) {
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Peer.java]: You're fiixing two different cases. I'd prefer to have them in two commits, one for each.

File: core/src/main/java/org/bitcoinj/core/Peer.java [hunk 3]
```diff
@@ -150,6 +150,8 @@ public GetDataRequest(Sha256Hash hash, SettableFuture future) {
     private final ReentrantLock lastPingTimesLock = new ReentrantLock();
     @GuardedBy("lastPingTimesLock") private long[] lastPingTimes = null;
     private final CopyOnWriteArrayList<PendingPing> pendingPings;
+    // Disconnect form a peer that is not responding to Pings
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Peer.java]: typo: form

Comment (pvyhnal-generalbytes) [reply core/src/main/java/org/bitcoinj/core/Peer.java]: Fixed, thanks

**Comment by schildbach** [comment]
I just merged the 'Ping' aspect of your PR. Can you rebase this PR on current master and squash everything into one commit? I need to evaluate the possible side effects of the 'pendingTxDownloads' aspect some more.

**Comment by pvyhnal-generalbytes** [comment]
I made a separate PR for the `pendingTxDownloads` https://github.com/bitcoinj/bitcoinj/pull/1734

**Comment by schildbach** [comment]
Btw. you don't need the create a new PR for rebases. Just force push (git push -f) your rebased brach over your old remote branch – github will pick it up automatically.

**Comment by pvyhnal-generalbytes** [comment]
Sorry, I wasn't sure about changing the commits that were already pushed, so I think we can close this PR and continue in #1734 
Just future reference the merged part of this PR is 29f450b8609fd17c33a2d6da5613fe63cb003601

**Comment by schildbach** [comment]
Closing this in favour of #1734.

**Pull request closed without merging**